tests/numerics/project_parallel_common.h

```diff
@@ -0,0 +1,359 @@
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/tria_boundary_lib.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/fe/fe_abf.h>
+#include <deal.II/fe/fe_dgp.h>
+#include <deal.II/fe/fe_dgp_monomial.h>
+#include <deal.II/fe/fe_dgp_nonparametric.h>
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_nedelec.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_q_hierarchical.h>
+#include <deal.II/fe/fe_raviart_thomas.h>
+#include <deal.II/fe/fe_system.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <fstream>
+#include <vector>
+
+
+template <int dim>
+void test ();
+
+
+template <int dim>
+class F :  public Function<dim>
+{
+public:
+  F (const unsigned int q,
+     const unsigned int n_components)
+    :
+    Function<dim>(n_components),
+    q(q)
+  {}
+
+  virtual double value (const Point<dim> &p,
+                        const unsigned int component) const
+  {
+    Assert ((component == 0) && (this->n_components == 1),
+            ExcInternalError());
+    double val = 0;
+    for (unsigned int d=0; d<dim; ++d)
+      for (unsigned int i=0; i<=q; ++i)
+        val += (d+1)*(i+1)*std::pow (p[d], 1.*i);
+    return val;
+  }
+
+
+  virtual void vector_value (const Point<dim> &p,
+                             Vector<double>   &v) const
+  {
+    for (unsigned int c=0; c<v.size(); ++c)
+      {
+        v(c) = 0;
+        for (unsigned int d=0; d<dim; ++d)
+          for (unsigned int i=0; i<=q; ++i)
+            v(c) += (d+1)*(i+1)*std::pow (p[d], 1.*i)+c;
+      }
+  }
+
+private:
+  const unsigned int q;
+};
+
+
+template <int dim, int components, int fe_degree>
+void do_project (const parallel::distributed::Triangulation<dim> &triangulation,
+                 const FiniteElement<dim> &fe,
+                 const unsigned int        order_difference)
+{
+  const unsigned int p = fe_degree;
+  DoFHandler<dim>        dof_handler(triangulation);
+  dof_handler.distribute_dofs (fe);
+
+  deallog << "n_dofs=" << dof_handler.n_dofs() << std::endl;
+
+  const MPI_Comm &mpi_communicator = triangulation.get_communicator();
+  const IndexSet locally_owned_dofs = dof_handler.locally_owned_dofs();
+  IndexSet locally_relevant_dofs;
+  DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+
+  ConstraintMatrix constraints;
+  constraints.reinit(locally_relevant_dofs);
+  DoFTools::make_hanging_node_constraints (dof_handler,
+                                           constraints);
+  constraints.close ();
+
+  TrilinosWrappers::MPI::Vector projection_tmp (locally_owned_dofs,
+                                                mpi_communicator);
+  TrilinosWrappers::MPI::Vector projection (locally_owned_dofs,
+                                            locally_relevant_dofs,
+                                            mpi_communicator);
```

Comment:
Is there a deeper reason to use Trilinos vectors here? You could also use `LinearAlgebra::distributed::Vector`, which would have the advantage that the test runs also without Trilinos.